src/cryptography/x509/extensions.py

```diff
@@ -108,17 +110,19 @@ def oid(self) -> ObjectIdentifier:
 
 
 class Extensions(object):
-    def __init__(self, extensions: typing.List["Extension"]):
+    def __init__(self, extensions: typing.List["Extension[ExtensionType]"]):
         self._extensions = extensions
 
-    def get_extension_for_oid(self, oid: ObjectIdentifier) -> "Extension":
+    def get_extension_for_oid(
+        self, oid: ObjectIdentifier
+    ) -> "Extension[ExtensionType]":
         for ext in self:
             if ext.oid == oid:
                 return ext
 
         raise ExtensionNotFound("No {} extension was found".format(oid), oid)
 
-    def get_extension_for_class(self, extclass) -> "Extension":
+    def get_extension_for_class(self, extclass) -> "Extension[ExtensionType]":
```

Comment:
Is there any way to type this so that it's known that `ExtensionType` here matches `extclass`?

Reply:
YES that is possible, like this:

```diff
diff --git a/src/cryptography/x509/extensions.py b/src/cryptography/x509/extensions.py
index 2858cef9..56f8b367 100644
--- a/src/cryptography/x509/extensions.py
+++ b/src/cryptography/x509/extensions.py
@@ -122,7 +122,9 @@ class Extensions(object):
 
         raise ExtensionNotFound("No {} extension was found".format(oid), oid)
 
-    def get_extension_for_class(self, extclass) -> "Extension[ExtensionType]":
+    def get_extension_for_class(
+        self, extclass: typing.Type[ExtensionType]
+    ) -> "Extension[ExtensionType]":
         if extclass is UnrecognizedExtension:
             raise TypeError(
                 "UnrecognizedExtension can't be used with "
```

but this makes mypy reveal another issue: 

1. `ExtensionType.oid` is defined as abstract property. Meaning for example `BasicConstraints(...).oid` works, but `BasicConstraints.oid` does **not**.
2. That's of course not actually true, as every subclass defines `oid` as a class attribute.
3. However `ExtensionNotFound` raised at the end of `get_extension_for_class` accesses it as a class attribute, triggering a mypy error.

Since the issue is not related to the change done by this PR, the function argument was not annotated before this PR either, and annotating it would reveal this issue even without my PR, I propose we ignore this issue here.

Reply:
PS: python has no `@classproperty` decorator, neither does the `abc` module. The solution I found was defining oid like this:

```diff
diff --git a/src/cryptography/x509/extensions.py b/src/cryptography/x509/extensions.py
index 91e9a30d..ec4ae9fb 100644
--- a/src/cryptography/x509/extensions.py
+++ b/src/cryptography/x509/extensions.py
@@ -102,11 +102,10 @@ class ExtensionNotFound(Exception):
 
 
 class ExtensionType(metaclass=abc.ABCMeta):
-    @abc.abstractproperty
-    def oid(self) -> ObjectIdentifier:
-        """
-        Returns the oid associated with the given extension type.
-        """
+    oid: typing.ClassVar[ObjectIdentifier]
+    """
+    Returns the oid associated with the given extension type.
+    """
 
 
 class Extensions(object):
```

... that works (test suite passes too). BUT since ExtensionType no longer has any abstract property (but maybe we could add/declare value to be an abstract property?), you now can actually create an ExtensionType instance.

Reply:
Thinking about this more closely, the diff from my initial comment won't do what you requested, it would only if this were a generic type parameter. I couldn't find anything to do that with concrete classes.

Not that the parameter shouldn't be annotated, but again I would do that in a separate PR, as it requires larger changes.